File: dashboard/deploy-vercel-api.mjs
#!/usr/bin/env node
/**
 * Deploy to Vercel using the REST API directly (no CLI needed).
 * Uses Node.js built-in fetch (Node 18+).
 */

import { readFileSync, readdirSync, statSync } from 'fs';
import { join, relative } from 'path';
import { execSync } from 'child_process';

const VERCEL_TOKEN = process.env.VERCEL_TOKEN || '1rzNjBUZLOAKORAXpSsZqEUI';
const API_BASE = 'https://api.vercel.com';

const headers = {
  'Authorization': `Bearer ${VERCEL_TOKEN}`,
  'Content-Type': 'application/json',
};

// Collect all files recursively, skipping node_modules, .next, .git
function collectFiles(dir, base = dir) {
  const files = [];
  for (const entry of readdirSync(dir)) {
    if (['node_modules', '.next', '.git', '.vercel', '__pycache__'].includes(entry)) continue;
    const fullPath = join(dir, entry);
    const stat = statSync(fullPath);
    if (stat.isDirectory()) {
      files.push(...collectFiles(fullPath, base));
    } else if (stat.size < 5 * 1024 * 1024) { // Skip files > 5MB
      const relativePath = relative(base, fullPath);
      const content = readFileSync(fullPath);
      files.push({
        file: relativePath,
        data: content.toString('base64'),
        encoding: 'base64',
      });
    }
  }
  return files;
}

async function deploy() {
  console.log('📦 Collecting dashboard files...');
  const projectDir = process.cwd();
  const files = collectFiles(projectDir);
  console.log(`Found ${files.length} files to deploy`);

  // Step 1: Create deployment
  console.log('🚀 Creating Vercel deployment...');
  const deployPayload = {
    name: 'solshield-dashboard',
    files: files,
    projectSettings: {
      framework: 'nextjs',
      buildCommand: 'npm run build',
      outputDirectory: '.next',
      installCommand: 'npm install',
    },
    target: 'production',
  };

  const response = await fetch(`${API_BASE}/v13/deployments`, {
    method: 'POST',
    headers,
    body: JSON.stringify(deployPayload),
  });

  const result = await response.json();

  if (!response.ok) {
    console.error('❌ Deployment failed:', JSON.stringify(result, null, 2));
    process.exit(1);
  }

  console.log('✅ Deployment created!');
  console.log(`📋 Deployment ID: ${result.id}`);
  console.log(`🌐 URL: https://${result.url}`);
  console.log(`🔗 Dashboard: https://vercel.com/dashboard`);
  
  // Poll for ready status
  console.log('\n⏳ Waiting for build to complete...');
  let status = result.readyState || result.status;
  let deployUrl = result.url;
  
  while (status === 'BUILDING' || status === 'INITIALIZING' || status === 'QUEUED') {
    await new Promise(r => setTimeout(r, 5000));
    const checkRes = await fetch(`${API_BASE}/v13/deployments/${result.id}`, { headers });
    const checkData = await checkRes.json();
    status = checkData.readyState || checkData.status;
    deployUrl = checkData.url || deployUrl;
    console.log(`  Status: ${status}...`);
  }

  if (status === 'READY') {
    console.log(`\n🎉 DEPLOYMENT LIVE!`);
    console.log(`🌐 Production URL: https://${deployUrl}`);
  } else {
    console.log(`\n⚠️ Final status: ${status}`);
    console.log(`🌐 URL: https://${deployUrl}`);
  }
}

deploy().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: dashboard/deploy-vercel-api.py
#!/usr/bin/env python3
"""Deploy dashboard to Vercel using the REST API with urllib (no curl needed)."""

import json
import os
import sys
import urllib.request
import urllib.error
import base64
import tarfile
import io
import glob

VERCEL_TOKEN = "1rzNjBUZLOAKORAXpSsZqEUI"
VERCEL_API = "https://api.vercel.com"

def read_file_bytes(path):
    with open(path, 'rb') as f:
        return f.read()

def collect_files(root_dir):
    """Collect all files for deployment, excluding unnecessary dirs."""
    files = []
    exclude_dirs = {'.next', 'node_modules', '.git', '__pycache__', '.vercel'}
    
    for dirpath, dirnames, filenames in os.walk(root_dir):
        dirnames[:] = [d for d in dirnames if d not in exclude_dirs]
        for filename in filenames:
            filepath = os.path.join(dirpath, filename)
            relpath = os.path.relpath(filepath, root_dir)
            try:
                content = read_file_bytes(filepath)
                files.append({
                    "file": relpath,
                    "data": base64.b64encode(content).decode('utf-8'),
                    "encoding": "base64"
                })
            except Exception as e:
                print(f"  Skipping {relpath}: {e}")
    return files

def api_request(endpoint, method="GET", data=None):
    """Make a Vercel API request."""
    url = f"{VERCEL_API}{endpoint}"
    headers = {
        "Authorization": f"Bearer {VERCEL_TOKEN}",
        "Content-Type": "application/json"
    }
    
    req = urllib.request.Request(url, headers=headers, method=method)
    if data:
        req.data = json.dumps(data).encode('utf-8')
    
    try:
        with urllib.request.urlopen(req, timeout=120) as resp:
            return json.loads(resp.read().decode('utf-8'))
    except urllib.error.HTTPError as e:
        body = e.read().decode('utf-8')
        print(f"API Error {e.code}: {body}")
        raise

def deploy():
    dashboard_dir = os.path.dirname(os.path.abspath(__file__))
    print(f"Deploying from: {dashboard_dir}")
    
    # Step 1: Collect files
    print("Collecting files...")
    files = collect_files(dashboard_dir)
    print(f"  Found {len(files)} files")
    
    if not files:
        print("ERROR: No files found!")
        sys.exit(1)
    
    # Step 2: Create deployment via Vercel API v13
    print("Creating Vercel deployment...")
    deploy_data = {
        "name": "solshield-dashboard",
        "files": files,
        "projectSettings": {
            "framework": "nextjs",
            "buildCommand": "npm run build",
            "outputDirectory": ".next",
            "installCommand": "npm install"
        },
        "target": "production"
    }
    
    result = api_request("/v13/deployments", method="POST", data=deploy_data)
    
    deploy_url = result.get("url", "unknown")
    deploy_id = result.get("id", "unknown")
    status = result.get("readyState", result.get("status", "unknown"))
    
    print(f"\n{'='*60}")
    print(f"🚀 DEPLOYMENT CREATED!")
    print(f"{'='*60}")
    print(f"  ID:     {deploy_id}")
    print(f"  URL:    https://{deploy_url}")
    print(f"  Status: {status}")
    print(f"{'='*60}")
    
    # Write URL to file for easy retrieval
    with open(os.path.join(dashboard_dir, ".vercel-url"), "w") as f:
        f.write(f"https://{deploy_url}")
    
    return deploy_url

if __name__ == "__main__":
    deploy()
